vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt

```diff
@@ -127,38 +127,81 @@ class VesselImpl(
         else -> DummyProfiler()
     }
 
-    private fun preloadImpl(entities: List<VesselEntity>) {
-        for (entity in entities) {
-            val kclass = try {
-                Class.forName(entity.type).kotlin
-            } catch (error: Exception) {
-                continue
+    /**
+     * Core preload implementation - is the same for blocking or suspend calls
+     * This is similar to what the code generated by Room does for both cases
+     *
+     * @return true only if the entire database could be loaded into cache in the allotted timeout
+     */
+    private fun preloadImpl(timeoutMS: Int?): Boolean {
+        val cursor = db.query("SELECT * FROM vessel", emptyArray())
+
+        cursor.use {
+            val typeCol = cursor.getColumnIndex("type")
+            val dataCol = cursor.getColumnIndex("data")
+
+            val startTimeMS = System.currentTimeMillis()
+            while (cursor.moveToNext()) {
+                val type = cursor.getString(typeCol)
+                val data = cursor.getString(dataCol)
+
+                val kclass = try {
+                    Class.forName(type).kotlin
+                } catch (error: Exception) {
+                    continue
+                }
+
+                val obj = if (data != null) fromJson(data, kclass) else nullValue
+                cache?.set(type, obj as Any)
+
+                timeoutMS?.let {
```

Comment:
I think we should try to leverage coroutine timeouts here instead - while the docs show cancellation is still asynchronous it will still happen earlier than waiting for the entire span like you have here.  https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-timeout.html

Reply:
I don't think that will work here:
>The code that is executing inside the [block](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-timeout.html) is cancelled on timeout and the active or next invocation of the cancellable suspending function inside the block throws a [TimeoutCancellationException](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-timeout-cancellation-exception/index.html).

There are no suspend points in this function, so it won't cancel.

Reply:
hmnm... maybe not... digging.
I'm a little worried about the timeout happening anywhere at anytime if that is indeed possible.  
I'd prefer to use this approach and have explicit control when the timeout is happening.